include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "include/proxy-wasm/compat.h"
+
+#include <time.h>
+#include <atomic>
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace proxy_wasm {
+
+#include "proxy_wasm_common.h"
+#include "proxy_wasm_enums.h"
+
+class WasmBase;
+class WasmVm;
+
+// TODO: use the SDK version when it is updated.
+enum class ProxyAction : uint32_t {
+  Illegal = 0,
+  Continue = 1,
+  Pause = 2,
+};
+
+using Pairs = std::vector<std::pair<string_view, string_view>>;
+using PairsWithStringValues = std::vector<std::pair<string_view, std::string>>;
+using CallOnThreadFunction = std::function<void(std::function<void()>)>;
+
+/**
+ * BufferInterface provides a interface between proxy-specifc buffers and the proxy-independent ABI
+ * implementation. Embedders should subclass BufferInterface to enable the proxy-independent code to
+ * implement ABI calls which use buffers (e.g. the HTTP body).
+ */
+struct BufferInterface {
+  virtual ~BufferInterface() {}
+  virtual size_t size() const = 0;
+  /**
+   * Copy bytes from the buffer into the Wasm VM corresponding to 'wasm'.
+   * @param start is the first buffer byte to copy.
+   * @param length is the length of sequence of buffer bytes to copy.
+   * @param ptr_ptr is the location in the VM address space to place the address of the newly
+   * allocated memory block which contains the copied bytes.
+   * @param size_ptr is the location in the VM address space to place the size of the newly
+   * allocated memory block which contains the copied bytes (e.g. length).
+   * @return true on success.
+   */
+  virtual bool copyTo(WasmBase *wasm, size_t start, size_t length, uint64_t ptr_ptr,
+                      uint64_t size_ptr) const = 0;
+};
+
+/**
+ * PluginBase is container to hold plugin information which is shared with all Context(s) created
+ * for a given plugin. Embedders may extend this class with additional host-specific plugin
+ * information as required.
+ */
+struct PluginBase {
+  PluginBase(string_view name, string_view root_id, string_view vm_id,
+             string_view plugin_configuration)
+      : name_(std::string(name)), root_id_(std::string(root_id)), vm_id_(std::string(vm_id)),
+        plugin_configuration_(plugin_configuration) {}
+
+  const std::string name_;
+  const std::string root_id_;
+  const std::string vm_id_;
+  std::string plugin_configuration_;
+  const std::string &log_prefix() const { return log_prefix_; }
+
+private:
+  std::string makeLogPrefix() const;
+
+  std::string log_prefix_;
+};
+
+/**
+ * ContextBase is the interface between the VM host and the VM. It has several uses:
```

Comment:
Can you add a few notes on why we have a single class for all possible ABI uses? It would be worth it to capture the design decisions here for future readers (our comment thread on this topic).

FWIW, after reading this code more, I think it's pretty confusing to have a giant class that implements everything. As a middle ground you might consider some pure virtual base classes for different concepts and then just have ContextBase derive from all of them. This would land you in the same situation you have today but would be more readable IMO.

Reply:
I will add some comments.  Unfortunately doing as you suggest for the onXXX functios would require replicating the full definitions in the inheriting class with an override since they are actually implemented there.  This (at least in part) defeats the purpose of clarity.

Reply:
> Unfortunately doing as you suggest for the onXXX functios would require replicating the full definitions in the inheriting class with an override since they are actually implemented there. This (at least in part) defeats the purpose of clarity.

Sorry I don't follow. If you ignore my point about pure virtual, and you are OK with multiple inheritance, can't you just have a bunch of base classes that are conceptually related that are split out from below and then derive from all of them in ContextBase? I think this would be way, way easier to understand for SDK writers. This class is going to keep growing and it will get even harder to understand. I'm confused why we would want this structure if the goal is to create an ecosystem of SDKs where we try to make it as easy as possible to write SDKs.

Reply:
Fine.  There will be duplication, but it will be easier to see than explain.  I'll pull them out into separate files.

Reply:
As you can see, it causes a lot of duplication and IMO is much more confusing because many of the calls out of the VM are shared between different types of plugin APIs and because other lifecycle calls are still on ContextBase.   We can't move the out calls and lifecycle calls into the individual interfaces without yet more duplication.  The end result is going to be much more complicated with lots of duplication and a maintenance buren because of all the duplication, and for what?  So that the user is forced to look at different files rather than simply read a block comment?  IMO this is far inferior to simply having some documentation for the integrator about what is required in order to bring up a new plugin type in a new environment.

Reply:
IMO this is an improvement. I actually don't think they even need to be in different files if you don't want that. They could be in this same file above. I also think that you can split out more methods such as all of the shared queue methods, gRPC methods, etc. We should get a second opinion though and if folks disagree with me that's fine, no problem. @htuch any thoughts here?

Reply:
I put them in different files because I for one will never read the _interface files.  I have to admit though that removing the big blocks of comments which largely violate the tenant of not commenting obvious things is a vast improvement, so perhaps something good will come out of this.  I have never though that javadoc style comments were of the slightest use, something which the Google C++ style team agrees with me on.   Nevertheless I would like to get this code upstreamed eventually, so on with the bikeshedding.

Reply:
>  I have to admit though that removing the big blocks of comments which largely violate the tenant of not commenting obvious things is a vast improvement, so perhaps something good will come out of this.

I would appreciate that we table the frustration. This may be obvious to you as you wrote all the code, but it's not obvious to anyone else. We need to have comments so that readers can understand the functions that they need to implement one way or the other. Whether they are "javadoc" style or "//" style I really don't care, but we need to have them somewhere. Thank you!

Reply:
FWIW, as someone fresh to this code, starting with the interfaces was super helpful in gaining an intuition about how things work at high level. I would strongly be in favor of this pattern. The comments in the interface file are super helpful.